crates/resolver-tests/src/lib.rs

```diff
@@ -227,51 +242,201 @@ fn sat_at_most_one(solver: &mut impl varisat::ExtendFormula, vars: &[varisat::Va
+
+                        solver.add_clause(&dep_clause);
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn process_pkg_dependencies(
+    solver: &mut varisat::Solver<'_>,
+    var_for_is_packages_used: &HashMap<PackageId, varisat::Var>,
+    var_for_is_packages_features_used: &HashMap<PackageId, HashMap<InternedString, varisat::Var>>,
+    pkg_var: varisat::Var,
+    pkg_dependencies: &[Dependency],
+    compatible_dep_summaries_by_name_in_toml: &IndexMap<InternedString, Vec<Summary>>,
+) {
+    for (dep, compatible_dep_summaries) in pkg_dependencies
+        .iter()
+        .zip(compatible_dep_summaries_by_name_in_toml.values())
+    {
+        // add clauses for package dependency features
+        for dep_summary in compatible_dep_summaries {
+            let dep_package_id = dep_summary.package_id();
+
+            let default_feature = if dep.uses_default_features()
+                && dep_summary.features().contains_key(&*INTERNED_DEFAULT)
+            {
+                Some(&INTERNED_DEFAULT)
+            } else {
+                None
+            };
+
+            for &feature_name in default_feature.into_iter().chain(dep.features()) {
+                solver.add_clause(&[
+                    pkg_var.negative(),
+                    var_for_is_packages_used[&dep_package_id].negative(),
+                    var_for_is_packages_features_used[&dep_package_id][&feature_name].positive(),
+                ]);
+            }
+        }
+
+        // active packages need to activate each of their non-optional dependencies
+        if !dep.is_optional() {
+            let dep_clause = compatible_dep_summaries
+                .iter()
+                .map(|d| var_for_is_packages_used[&d.package_id()].positive())
+                .chain([pkg_var.negative()])
+                .collect::<Vec<_>>();
+
+            solver.add_clause(&dep_clause);
+        }
+    }
+}
+
 /// Resolution can be reduced to the SAT problem. So this is an alternative implementation
 /// of the resolver that uses a SAT library for the hard work. This is intended to be easy to read,
 /// as compared to the real resolver.
 ///
-/// For the subset of functionality that are currently made by `registry_strategy` this will,
-/// find a valid resolution if one exists. The big thing that the real resolver does,
-/// that this one does not do is work with features and optional dependencies.
+/// For the subset of functionality that are currently made by `registry_strategy`,
+/// this will find a valid resolution if one exists.
 ///
-/// The SAT library dose not optimize for the newer version,
+/// The SAT library does not optimize for the newer version,
 /// so the selected packages may not match the real resolver.
-#[derive(Clone)]
-pub struct SatResolve(Rc<RefCell<SatResolveInner>>);
-
-struct SatResolveInner {
+pub struct SatResolver {
     solver: varisat::Solver<'static>,
+    old_root_vars: Vec<varisat::Var>,
```

Comment:
I don't think we need to keep `old_root_var` anymore. If cargo found a solution, then we make a assumption of everything we care about. If cargo did not find a solution, then SAT will opt out of as many constraints as it can including all the old `root_var` before deciding there is no solution.

So removing all code that refers to `old_root_vars` the code still passes all tests.

Reply:
The `old_root_vars` field is used to optimize the solving time. If we construct a `SatResolver` and then solve for a large number of root packages, adding assumptions that old root variables are negative allows the solver to completely ignore all previously added clauses, so it shouldn't take too much time to solve even after with a large number of total clauses.